components/style/gecko_selector_impl.rs

```diff
@@ -12,104 +12,130 @@ use stylesheets::Stylesheet;
+            }
+        }
+
+        PseudoElement(atom, is_anon_box)
+    }
+
+    #[inline]
+    fn from_atom(atom: &WeakAtom, in_ua: bool) -> Option<Self> {
+        use std::slice::Iter;
+        let mut slice = atom.as_slice();
+
+        let len = slice.len();
+        let mut iter = slice.iter();
+
+        if iter.next().map(|c| *c) != Some(':' as u16) {
+            return None
+        }
+
+        // This type is just needed because a plain iter.map() doesn't derive
+        // Clone because the closure is not Clone.
+        #[derive(Clone)]
+        struct Utf16toASCIIIter<'a>(Iter<'a, u16>);
+
+        impl<'a> Iterator for Utf16toASCIIIter<'a> {
+            type Item = u8;
+
+            fn next(&mut self) -> Option<u8> {
+                self.0.next().map(|c| {
+                    if *c <= 255 {
+                        *c as u8
+                    } else {
+                        // Something that can't exist in any pseudo-element.
+                        0u8
+                    }
+                })
+            }
+        }
+
+        Self::from_bytes_iter(Utf16toASCIIIter(iter), in_ua, len)
+    }
+
+    #[inline]
+    fn from_slice(s: &str, in_ua: bool) -> Option<Self> {
+        let bytes = s.as_bytes();
+        let len = bytes.len();
+        Self::from_bytes_iter(bytes.iter().cloned(), in_ua, len)
+    }
+
+    fn from_bytes_iter<I: Iterator<Item=u8> + Clone>(iter: I,
+                                                     in_ua_stylesheet: bool,
+                                                     byte_len: usize) -> Option<Self> {
+        use std::ascii::AsciiExt;
+        macro_rules! parse {
+            // NOTE: In an ideal world we should be able to either:
+            //
+            //  1. use only $pseudo_str, and generate the atom with
+            //     atom!(concat!(":", $pseudo_str))
+            //  2. use only the pseudo string with the colon, use
+            //     atom!($pseudo), and use $pseudo[1..] for comparing.
+            //
+            // This is not an ideal world, and none of those options work.
+            ($pseudo_str:expr, $atom:expr) => {
+                parse!($pseudo_str, $atom, false)
+            };
+            ($pseudo_str:expr, $atom:expr, $is_anon_box:expr) => {{
+                debug_assert!($pseudo_str[0..].chars().next().unwrap() != ':',
+                              "Parser doesn't provide semicolons");
+                debug_assert!($atom.chars().next().unwrap().unwrap()  == ':',
+                              "Gecko expects semicolons");
+                let bytes = $pseudo_str.as_bytes();
+                if byte_len == bytes.len() &&
+                    bytes.iter().zip(iter.clone()).all(|(a, b)| b.eq_ignore_ascii_case(a)) {
+                    return Some(PseudoElement($atom, $is_anon_box))
+                }
+            }}
+        }
+
+        parse!("before", atom!(":before"));
+        parse!("after", atom!(":after"));
```

Comment:
Do I understand correctly that this temporarily regresses our support for anon boxes during this patch? I don't see us passing |true| to parse! anywhere. I guess I'm ok with that, though generally would prefer to review less intermediate code...